pq/lms_sign_verify.c

```diff
@@ -0,0 +1,361 @@
+                          void * context);
+static int  lms_sign_verify(int levels, int height, int winternitz,
+                            size_t sigs_to_do);
+static void dump_hex(const char * what, const uint8_t * buf, size_t len);
+
+static WC_RNG rng;
+
+int
+main(int    argc,
+     char * argv[])
+{
+    int    levels = 0;
+    int    height = 0;
+    int    winternitz = 0;
+    size_t sigs_to_do = 1;
+    int    ret = 0;
+
+    if (argc < 4 || argc > 5) {
+        print_usage_and_die();
+    }
+
+    levels = atoi(argv[1]);
+    height = atoi(argv[2]);
+    winternitz = atoi(argv[3]);
+
+    if (argc == 5) {
+        sigs_to_do = atoll(argv[4]);
+    }
+
+    ret = wc_InitRng(&rng);
+    if (ret) {
+        fprintf(stderr, "error: wc_InitRng returned %d\n", ret);
+        return EXIT_FAILURE;
+    }
+
+    ret = lms_sign_verify(levels, height, winternitz, sigs_to_do);
+
+    wc_FreeRng(&rng);
+
+    return (ret == 0) ? EXIT_SUCCESS : EXIT_FAILURE;
+}
+
+static void
+print_usage_and_die(void)
+{
+    fprintf(stderr, "usage:\n");
+    fprintf(stderr, "  ./lms_sign_verify <levels> <height> <winternitz> [num signatures]\n");
+    fprintf(stderr, "\n");
+    fprintf(stderr, "examples:\n");
+    fprintf(stderr, "  ./lms_sign_verify 1 5 1\n");
+    fprintf(stderr, "  ./lms_sign_verify 3 5 4 100\n");
+    fprintf(stderr, "  ./lms_sign_verify 2 10 2 0\n");
+    fprintf(stderr, "\n");
+    fprintf(stderr, "description:\n");
+    fprintf(stderr, "  Generates an LMS/HSS key pair with L=levels, H=height, and\n");
+    fprintf(stderr, "  W=winternitz parameters, then signs and verifies a given\n");
+    fprintf(stderr, "  number of signatures (1 by default).\n");
+    fprintf(stderr, "\n");
+    fprintf(stderr, "  If 0 is given for num signatures, it prints the private and\n");
+    fprintf(stderr, "  public key as hex and exits early.\n");
+    fprintf(stderr, "\n");
+    fprintf(stderr, "notes:\n");
+    fprintf(stderr, " - The acceptable parameter values are those in RFC8554:\n");
+    fprintf(stderr, "     levels = {1..8}\n");
+    fprintf(stderr, "     height = {5, 10, 15, 20, 25}\n");
+    fprintf(stderr, "     winternitz = {1, 2, 4, 8}\n");
+    fprintf(stderr, "\n");
+    fprintf(stderr, " - The number of available signatures is\n");
+    fprintf(stderr, "     n = 2 ** (levels * height)\n");
+    fprintf(stderr, "\n");
+    fprintf(stderr, " - Larger winternitz values will reduce the signature size, at\n");
+    fprintf(stderr, "   the expense of longer key generation and signature times.\n");
+    fprintf(stderr, "\n");
+    fprintf(stderr, " - Key generation time is strongly determined by the height of\n");
+    fprintf(stderr, "   the first level tree. A 3 level, 5 height tree is much faster\n");
+    fprintf(stderr, "   than 1 level, 15 height at initial key gen, even if the number\n");
+    fprintf(stderr, "   of available signatures is the same.\n");
+
+    exit(EXIT_FAILURE);
+}
+
+static bool
```

Comment:
I don't think bool is always available. 
Make the callback return an integer and then define a new return code in wolfcrypt.  
The return code should be something explicit like SAVED_PRIVATE_KEY_TO_NV_MEMORY.

Reply:
The bools are gone. It returns a `WC_LMS_RC_*`  return code now.